### PR TITLE
Enhancement/add options support all data types

### PR DESCRIPTION
**Pull request by tamerharsh**
Please review.

### CODE REVIEW

File: commandlineparser/include/Parser.h [hunk 1]
```diff
@@ -62,6 +64,48 @@ namespace cmdParser
...
+			command_list.insert(KeyValue(obj->get_option_long_command(), obj));
+		}
+		else if (is_short_empty)
+		{
+			throw std::exception("Enter a valid short or long command or both.");
+		}
+	}
+
+	template<typename U>
+	std::vector <U> cmdParser::Parser::getValueAsString(const std::string & input) const
```
Comment (inbangsa) [inline commandlineparser/include/Parser.h]: change function name to get_value

File: commandlineparser/include/Parser.h [hunk 2]
```diff
@@ -62,6 +64,48 @@ namespace cmdParser
...
+
+	template <typename T>
+	void cmdParser::Parser::AddOptions(std::string short_command, std::string long_command, std::string short_description, std::string long_description, T def_value)
+	{
+		static_assert(!std::is_convertible<T, const char *>::value, "const char * data type for string literal is not supported in this version !");
+		
+		std::shared_ptr<cmdParser::Options> obj = std::make_shared <cmdParser::Options_with_value<T>>(short_command, long_command, short_description, long_description, def_value);
+
+		bool is_short_empty = false;
+
```
Comment (inbangsa) [inline commandlineparser/include/Parser.h]: Add a private function  add_option_objects and move the following code to that add function. 

private : 
add_option_object( std::shared_ptr<cmdParser::Options> obj)

File: commandlineparser/include/Parser.h [hunk 3]
```diff
@@ -62,6 +64,48 @@ namespace cmdParser
...
+	{
+		auto ptr = dynamic_cast<Options_with_value<U>*>(command_list.at(input).get());
+
+		if (ptr != nullptr)
+		{
+			return ptr->get_value();
+		}
+		else
+		{
+			throw std::exception(std::string(std::string("Type mismatch for input argument ") + input).c_str());
```
Comment (inbangsa) [inline commandlineparser/include/Parser.h]: Add a descriptive error message  saying what T is passed and if possible also tell  for current option what is the  T required

File: commandlineparser/src/Parser.cpp [hunk 4]
```diff
@@ -1,29 +1,5 @@
 #include "Parser.h"
```
Comment (inbangsa) [inline commandlineparser/src/Parser.cpp:2]: So this was was my point of  private constructor  delegating the other paramter along with  shared pointer<Options> obj

**Review by inbangsa — CHANGES_REQUESTED** [review]

File: commandlineparser/include/Parser.h (hunk 1 above)
Comment (tamerharsh) [reply commandlineparser/include/Parser.h]: done.

File: commandlineparser/include/Parser.h (hunk 2 above)
Comment (tamerharsh) [reply commandlineparser/include/Parser.h]: done.

File: commandlineparser/include/Parser.h (hunk 3 above)
Comment (tamerharsh) [reply commandlineparser/include/Parser.h]: Will do in next feature branch.  (#25 User should be able to get value as defined data types in C++ enhancement )

File: commandlineparser/src/Parser.cpp (hunk 4 above)
Comment (tamerharsh) [reply commandlineparser/src/Parser.cpp:2]: done .

**Pull request merged**